src/contract_client.rs

```diff
@@ -804,24 +800,208 @@ pub fn decode_concordium_std_error(reject_reason: i32) -> Option<ConcordiumStdRe
+/// were fitted with features to revert with different reasons (as defined in
+/// the smart contract logic). The `return_value` is used to distinguish between
+/// the different reasons for the revert coming from the smart contract logic.
+///
+/// There are some historical types used by the node which spread into this Rust
+/// SDK code base (such as `RejectReason`) that only include the `error_code`
+/// but not the `return_value`. The reason can be explained by the historical
+/// development of smart contracts on Concordium (types were not expanded to
+/// avoid breaking changes) as well as to keep the size of the node as small as
+/// possible since the `return_value` does not need to be saved by the node for
+/// achieving consensus.
+///
+/// As a result, this decoding function needs to have access to both the
+/// `return_value` and the `error_code` to decode the `reject_reason` of a
+/// reverted transaction into a human-readable error.
+///
+/// How is the `return_value` and `error_code` assigned in rejected
+/// transactions:
+/// - If the transaction reverts due to an error in the `concordium-std` crate,
+/// the `return_value` is None and the `error_code` is assigned as defined in
+/// the `concordium-std` crate.
+/// - If the transaction reverts due to an error in the smart contract logic:
+/// A smart contract V1 needs to implement a conversion to `Reject` for its
+/// smart contract errors.
+/// `<https://docs.rs/concordium-std/latest/concordium_std/struct.Reject.html>`
+///
+/// 1. Example: Deriving `Reject` in the smart contract.
+/// The simplest way to implement `Reject` in the smart contract is by deriving
+/// it.
+///
+/// ```ignore
+/// /// The custom errors the contract can produce.
+/// #[derive(Serialize, Debug, Reject, SchemaType)]
+/// pub enum CustomContractError {
+///     /// CustomError1 defined in the smart contract logic.
+///     CustomError1, // return_value: 00; error_code: -1
+///     /// CustomError2 defined in the smart contract logic.
+///     CustomError2, // return_value: 01; error_code: -2
+/// }
+/// ```
+///
+/// The `Reject` macro assigns the `error_codes` starting from `-1` to the enum
+/// variants and assigns the `return_values` by serializing the enum variants.
+/// The `return_values` are equivalent to the enum tags in the above example.
+///
+/// The JSON value returned by this function for the above `CustomError1` is:
+/// ```json
+/// {"CustomError1":[]}
+/// ```
+///
+/// 2. Example: Deriving `Reject` in the smart contract with nested errors.
+/// Nested errors are often used to inherit the errors from a smart
+/// contract library such as the cis2-library.
+/// `<https://github.com/Concordium/concordium-rust-smart-contracts/blob/dde42fa62254a55b46a4c9c52c32bbe661127001/concordium-cis2/src/lib.rs#L1093>`
+///
+/// Library:
+/// ```ignore
+/// /// The custom errors the library can produce.
+/// #[derive(Serialize, Debug, Reject, SchemaType)]
+/// pub enum Cis2Error<R> {
```

Comment:
I don't get this example.. For me it seems like you're implementing two custom errors in the same crate with one (`CustomContractError`) being nested in the other (`Cis2Error`).. It confuses me that `Cis2Error` does not have a reference to anything CIS-2 related? Wouldn't you normally have:

```rust
#[derive(Reject, SchemaType)]
pub enum CustomError {
  MyError,
  Cis2(concordium-cis2::Error) // Or whatever it's called
  ...
}
```

?

Reply:
I renamed the `Cis2Error` to `ParentError` since the example should be general.